### PR TITLE
Fix the redundant navigation error

### PR DESCRIPTION
**Pull request by absidue**
---
Fix the redundant navigation error
---

**Pull Request Type**

- [x] Bugfix

**Description**

Fixes the issue where spamming the nav buttons throws vue errors about redundant navigations.

**Screenshots (if appropriate)**
duplicate-navigation-error

**Testing (for code that is not small enough to be easily understandable)**
I spammed the nav buttons and the error didn't occur anymore.

**Desktop (please complete the following information):**
 - OS: Windows
 - OS Version: 10
 - FreeTube version: 7287e0011ee21cb6d827c979b4cff2a44bc2ee6b

**Additional context**

### CODE REVIEW

**Comment by ChunkyProgrammer** [comment]
Error still appears when clicking the profile setting icon
image

**Comment by ChunkyProgrammer** [comment]
Was able to have this error happen again when searching (search the same thing twice)

**Comment by vallode** [comment]
Potentially we need to hijack the prototype of the vue-router push function here? I mean is it feasible to expect every single next router interaction to always remember to put this if statement in place?

**Comment by absidue** [comment]
I thought about that but I haven't looked into the source code for that yet (if you want to take a look remember to look at the source code for the version that FreeTube uses). Although I'll need to expand the checks to also check the query parameters etc.

**Comment by absidue** [comment]
I've now removed the individual checks and wrapped the router push function in one that checks the path and query before navigating.

**Comment by PikachuEXE** [comment]
Can we have a summary of what to be tested?

**Review by PikachuEXE — COMMENTED** [review]
Tested
- spamming the nav buttons
- clicking the profile select's setting icon
- search the same thing twice
- Paste a YT Video URL to navbar, press arrow button multiple times

File: src/renderer/router/index.js [hunk 1]
```diff
@@ -14,9 +14,50 @@ import Playlist from '../views/Playlist/Playlist.vue'
...
+      }
+    } else {
+      if (
+        location.path !== router.currentRoute.path &&
+        (isQueryEmpty(location.query) ||
+          location.query !== router.currentRoute.query)
+      ) {
+        return super.push(location)
+      }
+    }
```
Comment (PikachuEXE) [inline src/renderer/router/index.js:54]:
```suggestion
    const currentQueryUSP = new URLSearchParams(router.currentRoute.query)
    let newPath = ''
    let newQueryUSP = new URLSearchParams()

    if (typeof location === 'string') {
      if (location.includes('?')) {
        const urlParts = location.split('?')
        newPath = urlParts[0]
        newQueryUSP = new URLSearchParams(urlParts[1])
      } else {
        newPath = location
        // newQueryUSP already empty
      }
    } else {
      newPath = location.path
      newQueryUSP = new URLSearchParams(location.query)
    }

    const pathsAreDiff = router.currentRoute.path !== newPath
    // Comparing `URLSearchParams` objects directly will always be different
    const queriesAreDiff = newQueryUSP.toString() !== currentQueryUSP.toString()

    if (pathsAreDiff || queriesAreDiff) {
      return super.push(location)
    }
```

**Review by PikachuEXE — CHANGES_REQUESTED** [review]
Cleanup duplicate statements (especially conditional)

`isQueryEmpty` can also be removed

**Review by PikachuEXE — APPROVED** [review]

File: src/renderer/router/index.js (hunk 1 above)
Comment (PikachuEXE) [reply src/renderer/router/index.js]: Irrelevant now but I think
```js
location.path !== router.currentRoute.path &&
        (isQueryEmpty(location.query) ||
          location.query !== router.currentRoute.query)
```
is a wrong conditional statement
Upper conditional statements are like (new path != cur path **or** new query != cur query) (I think)
This one is (new path != cur path **and** new query != cur query) ?

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
LGTM

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@ChunkyProgrammer sorry for the spam :D

**Pull request merged**